[See Toh Jin Wei] iP

seetohjinwei opened the pull request:
# Duke Application

> Best todo application ever! - Anonymous User

Duke is **great** for anyone who keeps a list of tasks! It's

* text-based
* simple to use
* ~FAST~ _SUPER FAST_ to run

All you need to do is,

- [ ] download the latest release from [here](https://github.com/seetohjinwei/ip/releases)!
- [ ] double-click it.
- [ ] add your tasks!
- [ ] ...
- [X] profit!

And it is **FREE**!

Features:

1.  Add tasks
1. Add deadlines
1. Add events
1. Delete them!
1. Find them by keywords!

This application is built in Java! Run `Duke` today!

```java
public static void main(String[] args) {
    new Duke().run();
}
```

<details>
<summary>SECRET :wink: </summary>
Please pretend this is a cool secret area!

Requirements:

- [X] a heading
- [X] a bullet list
- [X] a numbered list
- [X] a fenced code block (with syntax highlighting)
- [X] a task list
- [X] an emoji
- [X] a blockquote
- [X] a hyper link
- [X] inline code
- [X] some text formatting: bold, italic, stikethrough etc.
</details>

Commit:
Add Gradle support

seetohjinwei committed:
Add skeletal version of Duke

seetohjinwei committed:
Duke: add tasks

seetohjinwei committed:
Duke: list tasks

seetohjinwei committed:
Add Task class

seetohjinwei committed:
Duke class: change switch-case to conditional

seetohjinwei committed:
Task class: split toggleDone into markAsDone and markAsUndone

seetohjinwei committed:
Duke class: mark tasks as done and not done

seetohjinwei committed:
Add Todo class

Made Task class abstract

seetohjinwei committed:
Add Deadline class

seetohjinwei committed:
Add Event class

seetohjinwei committed:
Duke class: show number of tasks when adding

seetohjinwei committed:
Add automated text UI testing

seetohjinwei committed:
Add DukeException and error handling for Todo

Change "true" while loop to check for scanner input instead.

seetohjinwei committed:
Add error handling for Deadline

seetohjinwei committed:
Add error handling for Event

seetohjinwei committed:
Add error handling for unknown command

seetohjinwei committed:
Add tests for error handling

seetohjinwei committed:
Duke class: change task array to task arraylist

seetohjinwei committed:
Duke class: add ability to delete tasks

Add error handling and tests for deleting tasks.
Improve error handling for mark and unmark operations.

seetohjinwei committed:
Duke class: improved user input handling

seetohjinwei committed:
Improved input handling for various Todo classes

Abstracted out commonly used exceptions.

seetohjinwei committed:
Duke: Abstracted out operations into individual methods

seetohjinwei committed:
Load tasks from file

seetohjinwei committed:
Save tasks to file

seetohjinwei committed:
Change dates from String to LocalDate

seetohjinwei committed:
Merge branch 'branch-Level-7'

seetohjinwei committed:
Merge branch 'branch-Level-8'

seetohjinwei committed:
Fix bug on reading saved dates

seetohjinwei committed:
Add Storage class

seetohjinwei committed:
Add Ui class

seetohjinwei committed:
Add Command, CommandResult and *Command classes

seetohjinwei committed:
Add Parser class

seetohjinwei committed:
Add TaskList class

seetohjinwei committed:
Reorganise Duke class

seetohjinwei committed:
Divide classes into packages

seetohjinwei committed:
Add Gradle for build, run, test

seetohjinwei committed:
Add some JUnit tests

seetohjinwei committed:
Add shadowjar plugin to Gradle

seetohjinwei committed:
Add JavaDoc comments

seetohjinwei committed:
Follow coding standard

seetohjinwei committed:
Find task by keyword

seetohjinwei committed:
Merge branch 'branch-A-JavaDoc'

seetohjinwei committed:
Merge branch 'branch-A-CodingStandard'

seetohjinwei committed:
Fix some constant names

seetohjinwei committed:
Merge branch 'branch-Level-9'

seetohjinwei committed:
Add no tasks found message

seetohjinwei committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

seetohjinwei committed:
Change LocalDate to LocalDateTime

seetohjinwei committed:
Fix minor coding standard violations

Remove unnecessary this keyword.
Use @inheritDoc.

seetohjinwei committed:
Add basic JavaFX GUI

Also, remove text based UI and tests

seetohjinwei committed:
Add command history

Navigate with up and down arrow keys

seetohjinwei committed:
Merge branch 'branch-Level-10'

seetohjinwei committed:
Minor change to variable name

TanPingZhi commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class FindCommand extends Command {
+    /** Command word of the find command. */
+    public static final String COMMAND_WORD = "find";
+    private static final String TASK_FORMAT = "%d: %s";
+    private static final String USER_MESSAGE_FORMAT = "Here are the matching tasks in your list!\n%s";
+    private static final DukeException WRONG_FORMAT =
+        new DukeException("Wrong format for Find!\nShould be 'find <keyword>'.");
+    private final String keyword;
```
Good use of constants

seetohjinwei replied:
Thanks for taking the time to provide your feedback!

TanPingZhi reviewed (commented):
Nice
No issues with naming or layouts

laxus2308 reviewed (commented):
I like how u provide additional comments for your variables and code to help me understand better. I like the naming of methods, especially the boolean methods which are really clear what the method is for. The imports are also grouped together for related imports, which is something I like as well. Great use of constants to reduce duplicate code. Very consistent format of code. I like the comprehensive test cases as well. Could not spot any coding standard violations throughout. Well Done!

seetohjinwei committed:
Add assertions

Code lacks assertions.

Add assertions to ensure that certain conditions are met.

Let's ensure that tasks are set before executing command.

seetohjinwei committed:
Improve code quality

Some code is too complex.

Reduce code complexity, mainly through abstracting
long or heavily nested methods.

Let's ensure that code quality is kept up.

seetohjinwei committed:
Add streams

Find command's execution is confusing.

Use streams to improve readability of the code.

Let's ensure that streams are used to simplify code.

seetohjinwei committed:
Merge pull request #1 from seetohjinwei/branch-A-Assertions

Add assertions

seetohjinwei committed:
Merge branch 'master' into branch-A-CodeQuality

seetohjinwei committed:
Merge pull request #2 from seetohjinwei/branch-A-CodeQuality

Improve code quality

seetohjinwei committed:
Merge branch 'master' into branch-A-Streams

seetohjinwei committed:
Merge pull request #3 from seetohjinwei/branch-A-Streams

Add streams

seetohjinwei committed:
Add undo command

There is no way to undo commands.

Add undo command to undo commands.

Let's support undoing actions.

seetohjinwei committed:
Merge branch 'branch-C-Undo'

seetohjinwei committed:
Add continuous integration

Charles1026 commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+public class DeleteCommand extends Command {
+    /** Command word of the delete command. */
+    public static final String COMMAND_WORD = "delete";
+    private static final String USER_MESSAGE_FORMAT = "Removing this task!\n  %s\nNow you have %d tasks left.";
```
Smart use of formatting 👍

Charles1026 commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,24 @@
+ * Represents an exception in the Duke application.
+ */
+public class DukeException extends Exception {
+    /** Exception due to an invalid index. */
```
Can extend these into individual exception classes

Charles1026 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,130 @@
+                tasks.addTask(task);
+            }
+        } catch (FileNotFoundException e) {
+            // Should not happen because file is created beforehand.
```
Creating file could fail sometimes

Charles1026 reviewed (commented):
👍

seetohjinwei committed:
Organise exceptions better

seetohjinwei committed:
Fix some comments and SLAP

seetohjinwei committed:
Update user guide

seetohjinwei committed:
Fix undo actions order

seetohjinwei committed:
Add message when there are no tasks to list

seetohjinwei committed:
Add resizing to GUI

seetohjinwei committed:
Improve styling of GUI

seetohjinwei committed:
Improve messages

- Add welcome message
- Add personality
- Add bye message with delay before exit
- Ignore empty command
- Add application title

seetohjinwei committed:
Add acknowledgement

seetohjinwei committed:
Update README.md

seetohjinwei committed:
Fix for Linux

seetohjinwei committed:
GUI: Set minimum height and width of stage

seetohjinwei committed:
Fix crash on Windows and Linux
